prov/efa/src/rxr/rxr_ep.c

```diff
@@ -2502,14 +2501,19 @@ int rxr_ep_alloc_app_device_info(struct fi_info **app_device_info,
 }
 
 /**
- * @brief Create ibv_cq_ex by calling ibv_create_cq_ex
+ * @brief Set rxr_ep->ibv_cq_ex by calling efadv_create_cq or ibv_create_cq_ex
  *
- * @param[in] cq Pointer to the efa_cq.
- * @param[in] attr Pointer to fi_cq_attr.
- * @return Return pointer to ibv_cq_ex if successful, otherwise NULL
+ * @param[in,out] ep Pointer to rxr_ep
+ * @param[in] attr Pointer to fi_cq_attr
+ * @return Return 0 on success, error code otherwise
  */
-static inline struct ibv_cq_ex *rxr_ep_create_ibv_cq_ex(struct rxr_ep *ep, struct fi_cq_attr *attr)
+#if HAVE_EFADV_CQ_EX
+static inline int rxr_ep_create_ibv_cq_ex(struct rxr_ep *ep, struct fi_cq_attr *attr)
```

Comment:
These two `rxr_ep_create_ibv_cq_ex` has considerable code duplication. Consider how to reduce it.

Reply:
That's why I originally had `#if HAVE_EFADV_CQ_EX` inside the function but you said the libfabric convention is to have separate functions. The other way to reduce duplication is to move some code to `rxr_endpoint` but that would make the already long `rxr_endpoint` longer. Any ideas on how exactly to reduce the duplication?

Reply:
You need to find the minimum section of code that has branching for both pre-processors, make them as a function, then use the same function to call it.